File: lib/polymer_element.dart
@JS('Polymer')
library polymer_element;

import 'package:html5/html.dart';
import 'dart:js';
import 'package:js/js.dart';
import 'package:js/js_util.dart';
export 'super.dart' show callSuper;
export 'package:polymer_element/annotations.dart';
import 'package:polymer_element/annotations.dart';
// ignore: UNUSED_IMPORT
import 'package:polymer_element/polymerize_js.dart';
// ignore: UNUSED_IMPORT
import 'metadata_registry.dart';
import 'package:polymerize_common/init.dart';
export 'metadata_registry.dart';

const String POLYMER_VERSION = "v2.0.0";
const String WEB_COMPONENTS = "v1.0.0";

class Config {
  List observers;

  Config({this.observers});
}



const _Undefined = const {};

class EventOptions {
  final bool bubbles;
  final bool cancelable;
  final HTMLElement node;

  const EventOptions({this.bubbles: true, this.cancelable: false, this.node});
}

Event createCustomEvent(String type,
    [detail, EventOptions opt = const EventOptions()]) {
  Event ev = new CustomEvent(
      type,
      new CustomEventInit()
        ..bubbles = opt.bubbles
        ..cancelable = opt.cancelable
        ..detail = detail);
  return ev;
}

getDetail(Event ev) => (new JsObject.fromBrowserObject(ev))['detail'];

@JS('DomRepeat')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/elements/dom-repeat.html',
    name: 'polymer')
@PolymerRegister('dom-repeat', native: true)
abstract class DomRepeat implements PolymerElement {
  external itemForElement(el);
  external indexForElement(el);
}

@JS('DomIf')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/elements/dom-if.html',
    name: 'polymer')
@PolymerRegister('dom-if', native: true)
abstract class DomIf implements PolymerElement {
}

@JS('Templatizer')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/legacy/templatizer-behavior.html',
    name: 'polymer')
abstract class Templatizer {
  external static flush();
  external PolymerElement templatize(HTMLTemplateElement template, options);
}

@JS('PropertyEffects')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/mixins/property-effects.html',
    name: 'polymer')
abstract class PropertyEffects {
  external notifyPath(String path, [val = _Undefined]);

  external push(String path, item);

  external pop(String path);

  external shift(String path, vals);

  external unshift(String path, item);

  external get(String path, [root]);

  external set(String path, value, [root]);

  external linkPaths(String from, String to);

  external unlinkPaths(String to);

  external notifySplice(String path, splices);
}

@JS('TemplateInstanceBase')
abstract class TemplateInstanceBase implements PropertyEffects {
  external forwardHostProp(prop, value);
  external get parentModel;
}

@JS()
@anonymous
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/utils/templatize.html',
    name: 'polymer')
abstract class TemplatizeModule {
  external TemplateInstanceBase templatize(
      HTMLTemplateElement template, owner, options);
  external modelForElement(host, Element element);
}

@JS('Templatize')
external TemplatizeModule get Templatize;

/**
 * Polymer hybrid behavior in order to mark the element
 * as a mutable data.
 */
@JS('MutableDataBehavior')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/legacy/mutable-data-behavior.html',
    name: 'polymer')
abstract class MutableDataBehavior {}

@JS('OptionalMutableDataBehavior')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/legacy/mutable-data-behavior.html',
    name: 'polymer')
abstract class OptionalMutableDataBehavior {}

@JS('MutableData')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/mixins/mutable-data.html',
    name: 'polymer')
abstract class MutableData {}

@JS('OptionalMutableData')
@BowerImport(
    ref: POLYMER_VERSION,
    import: 'polymer/lib/mixins/mutable-data.html',
    name: 'polymer')
abstract class OptionalMutableData {}

@JS('ElementMixin')
@BowerImport(
    ref: POLYMER_VERSION,
    import: "polymer/lib/mixins/element-mixin.html",
    name: 'polymer')
abstract class ElementMixin implements PropertyEffects {
  external get $;

  external $$(String selector);
}

@JS('Element')
@BowerImport(
    ref: POLYMER_VERSION, import: "polymer/polymer.html", name: 'polymer')
abstract class PolymerElement implements HTMLElement, ElementMixin {
  external get $;

  external $$(String selector);

  external ready();

  external connectedCallback();

  external disconnectedCallback();

  external attributeChangedCallback(name, old, value);
}


/**
 * Unfortunately JS-INTEROP doesn't allows for varargs or
 * more complex method mapping, so we have to provide those
 * ugly static utility methods.
 */
class PropertyEffectsUtils {
  static splice(PropertyEffects el, String path, int index, int howmany,
          List items) =>
      callMethod(
          el as dynamic, 'splice', [path, index, howmany]..addAll(items));
  static unshift(PropertyEffects el, String path, [items]) =>
      callMethod(el as dynamic, 'unshift', [path]..addAll(items));

  static push(PropertyEffects el, String path, [items]) =>
      callMethod(el as dynamic, 'push', [path]..addAll(items));
}

File: lib/polymer_element.ts
/** Library asset:polymer_element/lib/polymer_element.dart */
import {is,isNot,equals} from "@dart2ts/dart/_common";
import {defaultConstructor,namedConstructor,namedFactory,defaultFactory,DartClass,Implements,With,op,Op,OperatorMethods,DartClassAnnotation,DartMethodAnnotation,DartPropertyAnnotation,Abstract,AbstractProperty,int,bool,double,Omit} from "@dart2ts/dart/utils";
import * as _common from "@dart2ts/dart/_common";
import * as core from "@dart2ts/dart/core";
import * as async from "@dart2ts/dart/async";
import * as lib3 from "@dart2ts.packages/html5/lib/html";
import * as js_util from "@dart2ts/dart/js_util";

export var createCustomEvent : (type : string,detail? : any,opt? : EventOptions) => Event = (type : string,detail? : any,opt? : EventOptions) : Event =>  {
    opt = opt || new EventOptions();
    let ev : Event = new CustomEvent(type,((_) : CustomEventInit =>  {
        {
            _.bubbles = opt.bubbles;
            _.cancelable = opt.cancelable;
            _.composed = opt.composed;
            _.detail = detail;
            return _;
        }
    })({
    } as CustomEventInit));
    return ev;
};
/*
export var stamp : (templateClass : any,model? : core.DartList<any>) => Polymer.TemplateInstance = (templateClass : any,model? : core.DartList<any>) : Polymer.TemplateInstance =>  {
    return js_util.callConstructor(templateClass,model);
};
*/
export namespace EventOptions {
    export type Constructors = 'EventOptions';
    export type Interface = Omit<EventOptions, Constructors>;
}
@DartClass
export class EventOptions {
    bubbles : boolean;

    cancelable : boolean;

    composed : boolean;

    node : HTMLElement;

    constructor(_namedArguments? : {bubbles? : boolean,cancelable? : boolean,node? : HTMLElement,composed? : boolean}) {
    }
    @defaultConstructor
    EventOptions(_namedArguments? : {bubbles? : boolean,cancelable? : boolean,node? : HTMLElement,composed? : boolean}) {
        let {bubbles,cancelable,node,composed} = Object.assign({
            "bubbles" : true,
            "cancelable" : false,
            "composed" : true}, _namedArguments );
        this.bubbles = bubbles;
        this.cancelable = cancelable;
        this.node = node;
        this.composed = composed;
    }
}
/*
export namespace PropertyEffectsUtils {
    export type Constructors = 'PropertyEffectsUtils';
    export type Interface = Omit<PropertyEffectsUtils, Constructors>;
}
@DartClass
export class PropertyEffectsUtils {
    static splice(el : Polymer.PropertyEffects,path : string,index : number,howmany : number,items : core.DartList<any>) {
        return js_util.callMethod(el as any,'splice',((_) : core.DartList<any> =>  {
            {
                _.addAll(items);
                return _;
            }
        })(new core.DartList.literal(path,index,howmany)));
    }
    static unshift(el : Polymer.PropertyEffects,path : string,items? : any) {
        return js_util.callMethod(el as any,'unshift',((_) : core.DartList<any> =>  {
            {
                _.addAll(items);
                return _;
            }
        })(new core.DartList.literal(path)));
    }
    static push(el : Polymer.PropertyEffects,path : string,items? : any) {
        return js_util.callMethod(el as any,'push',((_) : core.DartList<any> =>  {
            {
                _.addAll(items);
                return _;
            }
        })(new core.DartList.literal(path)));
    }
    constructor() {
    }
    @defaultConstructor
    PropertyEffectsUtils() {
    }
}
*/
export class properties {
    private static __$POLYMER_VERSION : string;
    static get POLYMER_VERSION() : string { 
        if (this.__$POLYMER_VERSION===undefined) {
            this.__$POLYMER_VERSION = "v2.0.1";
        }
        return this.__$POLYMER_VERSION;
    }
    static set POLYMER_VERSION(__$value : string)  { 
        this.__$POLYMER_VERSION = __$value;
    }

    private static __$WEB_COMPONENTS : string;
    static get WEB_COMPONENTS() : string { 
        if (this.__$WEB_COMPONENTS===undefined) {
            this.__$WEB_COMPONENTS = "v1.0.1";
        }
        return this.__$WEB_COMPONENTS;
    }
    static set WEB_COMPONENTS(__$value : string)  { 
        this.__$WEB_COMPONENTS = __$value;
    }

    private static __$_Undefined;
    static get _Undefined() { 
        if (this.__$_Undefined===undefined) {
            this.__$_Undefined = new core.DartMap.literal([
            ]);
        }
        return this.__$_Undefined;
    }
    static set _Undefined(__$value : any)  { 
        this.__$_Undefined = __$value;
    }

}
